lib/types.d.ts

```diff
@@ -0,0 +1,204 @@
+ *
+ *     // Collection Keys
+ *     COLLECTION: {
+ *         REPORT: 'report_',
+ *     },
+ * } as const;
+ *
+ * type OnyxKeysMap = typeof ONYXKEYS;
+ * type OnyxCollectionKey = ValueOf<OnyxKeysMap['COLLECTION']>;
+ * type OnyxKey = DeepValueOf<Omit<OnyxKeysMap, 'COLLECTION'>>;
+ *
+ * type OnyxValues = {
+ *     [ONYXKEYS.ACCOUNT]: Account;
+ *     [ONYXKEYS.IS_SIDEBAR_LOADED]: boolean;
+ *     [ONYXKEYS.COLLECTION.REPORT]: Report;
+ * };
+ *
+ * export default ONYXKEYS;
+ * export type {OnyxKey, OnyxCollectionKey, OnyxValues};
+ *
+ * // global.d.ts
+ * import {OnyxKey, OnyxCollectionKey, OnyxValues} from './ONYXKEYS';
+ *
+ * declare module 'react-native-onyx' {
+ *     interface CustomTypeOptions {
+ *         keys: OnyxKey;
+ *         collectionKeys: OnyxCollectionKey;
+ *         values: OnyxValues;
+ *     }
+ * }
+ * ```
+ */
+interface CustomTypeOptions {}
+
+/**
+ * Represents a string union of all Onyx normal keys.
+ */
+type Key = TypeOptions['keys'];
+
+/**
+ * Represents a string union of all Onyx collection keys.
+ */
+type CollectionKeyBase = TypeOptions['collectionKeys'];
+
+/**
+ * Represents a literal string union of all Onyx collection keys.
+ * It allows appending a string after each collection key e.g. `report_some-id`.
+ */
+type CollectionKey = `${CollectionKeyBase}${string}`;
+
+/**
+ * Represents a string union of all Onyx normal and collection keys.
+ */
+type OnyxKey = Key | CollectionKey;
+
+/**
+ * Represents a mapping of Onyx keys to values, where keys are either normal or collection Onyx keys
+ * and values are the corresponding values in Onyx's state.
+ *
+ * For collection keys, `KeyValueMapping` allows any string to be appended
+ * to the key (e.g., 'report_some-id', 'download_some-id').
+ *
+ * The mapping is derived from the `values` property of the `TypeOptions` type.
+ */
+type KeyValueMapping = {
+    [TKey in keyof TypeOptions['values'] as TKey extends CollectionKeyBase
+        ? `${TKey}${string}`
+        : TKey]: TypeOptions['values'][TKey];
+};
+
+/**
+ * Represents a selector function type which operates based on the provided `TKey` and `ReturnType`.
+ *
+ * A `Selector` is a function that accepts a value and returns a processed value.
+ * This type accepts two type parameters: `TKey` and `TReturnType`.
+ *
+ * The type `TKey` extends `OnyxKey` and it is the key used to access a value in `KeyValueMapping`.
+ * `TReturnType` is the type of the returned value from the selector function.
+ *
+ * If `KeyValueMapping[TKey]` is equal to 'unknown', the `Selector` type represents a function
+ * that takes any value and returns an `unknown` value.
+ *
+ * If `KeyValueMapping[TKey]` is not 'unknown', the `Selector` type represents a function that
+ * takes either the value of type `OnyxEntry<KeyValueMapping[TKey]>`, and returns a value of `TReturnType`.
+ */
+type Selector<TKey extends OnyxKey, TReturnType> = IsEqual<KeyValueMapping[TKey], unknown> extends true
```

Comment:
In what case `KeyValueMapping[TKey]` be `unknown`?

I think you meant `IsEqual<KeyValueMapping[TKey], TReturnType>`?

Reply:
why is this know just

```ts
type Selector<TKey extends OnyxKey, TReturnType> = (value: OnyxEntry<KeyValueMapping[TKey]>) => TReturnType;
```

Reply:
I think the logic is correct, `unknown` is a default type when no type is set for a key in `OnyxValues`. Selector is `(value: unknown) => unknown` in such case. @fabioh8010 wdyt?

Reply:
@hayata-suenaga @blazejkustra I was doing some testing here and I think this is not necessary anymore, just kept 

```ts
type Selector<TKey extends OnyxKey, TReturnType> = (value: OnyxEntry<KeyValueMapping[TKey]>) => TReturnType;
```

and things seems to be working the same whether the project augmented Onyx or not. Will change it.

Reply:
@hayata-suenaga @blazejkustra  Adjusted!